runtime/lib/ttnn/program.cpp

```diff
@@ -278,16 +278,26 @@ run(::tt::target::ttnn::TransposeOp const *op, ::ttnn::device::Device &device,
   int32_t dimension1 = op->dimension1();
   int32_t dimension2 = op->dimension2();
   auto input_rank = in.get_shape().rank();
-  std::vector<std::int64_t> dimensionOrder(input_rank);
+  // for the current version of permute, we need to work in 4D, so we add
+  // leading dimensions of size 1
+  std::vector<std::int64_t> dimensionOrder(4);
   std::iota(dimensionOrder.begin(), dimensionOrder.end(), 0);
   if (dimension1 < 0) {
-    dimension1 += input_rank;
+    dimension1 += 4;
+  } else {
+    dimension1 = dimension1 + 4 - input_rank;
   }
+
   if (dimension2 < 0) {
-    dimension2 += input_rank;
+    dimension2 += 4;
+  } else {
+    dimension2 = dimension2 + 4 - input_rank;
   }
   std::swap(dimensionOrder[dimension1], dimensionOrder[dimension2]);
-  tensorPool.push_back(::ttnn::permute(in, dimensionOrder));
+  // Ideally this would use ttnn::transpose, but since ttnn::transpose doesn't
+  // work at the moment, we use this temporary solution.
+  auto unsqueezed_input = ::ttnn::unsqueeze_to_4D(in);
```

Comment:
I think we should push this also to tensor pool, to avoid race conditions when `unsqueezeed_input` is deleted before `permute` is executed...

Reply:
I _think_ ttnn will bump the reference count to this intermediate tensor and keep it alive until the op has finished executing.  Probably worth following up with ttnn folks to confirm, but this would be ideal.

The tensor pool is intended to keep tensors alive between ops, so they can be passed to each other, so pending my statement above the code should stay the way it reads so that we don't have to manually delete the temporary.

Reply:
These kinds of temporaries are tricky though, I'm not sure if `unsqueeze_to_4D` actually allocates anything or not (probably not).  But if it did, we'd have to tell the compiler about this so it can account for there being an intermediate op.  Out of scope for this review though

Reply:
> I think ttnn will bump the reference count to this intermediate tensor and keep it alive until the op has finished executing. Probably worth following up with ttnn folks to confirm, but this would be ideal.

yeah, this seems to be the case